tensorflow_addons/image/filters.py

```diff
@@ -34,82 +35,88 @@ def func2():
+
+        rank = image.shape.rank
+        if rank != 3 and rank != 4:
+            raise ValueError("image should be either 3 or 4-dimensional.")
+
+        if padding not in ["REFLECT", "CONSTANT", "SYMMETRIC"]:
             raise ValueError(
-                'Number of Pixels in each dimension of the image should be \
-                more than the filter size. Got filter_shape (%sx' %
-                filter_shape[0] + '%s).' % filter_shape[1] +
-                ' Image Shape (%s)' % image.shape)
-        if filter_shapex % 2 == 0 or filter_shapey % 2 == 0:
-            raise ValueError('Filter size should be odd. Got filter_shape (%sx'
-                             % filter_shape[0] + '%s)' % filter_shape[1])
-        image = tf.cast(image, tf.float32)
-        tf_i = tf.reshape(image, [row * col * ch])
-        ma = tf.math.reduce_max(tf_i)
-        image = _normalize(image, ma)
+                "padding should be one of \"REFLECT\", \"CONSTANT\", or "
+                "\"SYMMETRIC\".")
 
-        # k and l is the Zero-padding size
+        filter_shape = keras_utils.conv_utils.normalize_tuple(
+            filter_shape, 2, "filter_shape")
 
-        listi = []
-        for a in range(ch):
-            img = image[:, :, a:a + 1]
-            img = tf.reshape(img, [1, row, col, 1])
-            slic = tf.image.extract_patches(
-                img, [1, filter_shapex, filter_shapey, 1], [1, 1, 1, 1],
-                [1, 1, 1, 1],
-                padding='SAME')
-            li = tf.reduce_mean(slic, axis=-1)
-            li = tf.reshape(li, [row, col, 1])
-            listi.append(li)
-        y = tf.concat(listi[0], 2)
+        # Expand to a 4-D tensor
+        if rank == 3:
+            image = tf.expand_dims(image, axis=0)
 
-        for i in range(len(listi) - 1):
-            y = tf.concat([y, listi[i + 1]], 2)
+        # Keep the precision if it's float;
+        # otherwise, convert to float32 for computing.
+        orig_dtype = image.dtype
+        if not image.dtype.is_floating:
+            image = tf.dtypes.cast(image, tf.dtypes.float32)
 
-        y *= 255
-        y = tf.cast(y, tf.int32)
+        # Explicitly pad the image
+        image = _pad(
+            image, filter_shape, mode=padding, constant_values=constant_values)
 
-        return y
+        # Filter of shape (filter_width, filter_height, in_channels, 1)
+        # has value 1 / (filter_width * filter_height) for each element.
+        area = filter_shape[0] * filter_shape[1]
+        filter_shape = filter_shape + (tf.shape(image)[-1], 1)
+        kernel = tf.ones(shape=filter_shape, dtype=image.dtype) / area
+
+        output = tf.nn.depthwise_conv2d(
+            image, kernel, strides=(1, 1, 1, 1), padding="VALID")
+
+        # Squeeze out the first axis to make sure
+        # output has the same dimension with image.
+        if rank == 3:
+            output = tf.squeeze(output, axis=0)
+
+        return tf.dtypes.cast(output, orig_dtype)
```

Comment:
I'm afraid we should let users do the conversion if they'd like, because the operation might cause loss of precision (say, uint8)

Reply:
Do you mean that we should return the float image? I'm fine with that though.

Reply:
Let's return float image. And add an argument later if anyone requests it. What do you think :-)

Reply:
If float image is retuned, the floating range should be in between 0.0 and 1.0. Because many image showing tools like matplotlib can not create the image properly if the values are in the range 0.0 to 255.0. If you wish to return between 0 to 255 range,  i think int32 will be a better choice.  In the original code int range was used.  I also think there should be uniformity in the output type.

Reply:
I have two concerns here:

1. Since median operation might not require explicitly casting, `median_filter2d` would be suitable to return the image of original dtype. This would result in the mismatched API for `mean_filter2d` and `median_filter2d`.

2. Scipy returns the image of original dtype

Reply:
We can explicitly cast the both filter operation to same type.  Let it be int32 or float64 or anything else.  That will only ensure uniformity in output.

Reply:
OK, let's keep it simple, and don't return float image. I'm fine with the current implementation: returns the image of original dtype.